pageserver/src/page_service.rs

```diff
@@ -738,6 +741,58 @@ impl postgres_backend::Handler for PageServerHandler {
 
             pgb.write_message_noflush(&SINGLE_COL_ROWDESC)?
                 .write_message_noflush(&BeMessage::CommandComplete(b"SELECT 1"))?;
+        } else if query_string.starts_with("get_lsn_by_xid ") {
+            // Locate LSN of transaction with given XID
+            // TODO lazy static
+            let re = Regex::new(r"^get_lsn_by_xid ([[:xdigit:]]+) ([[:xdigit:]]+) ([[:digit:]]+)$")
+                .unwrap();
+            let caps = re
+                .captures(query_string)
+                .with_context(|| format!("invalid get_lsn_by_xid: '{}'", query_string))?;
+
+            let tenantid = ZTenantId::from_str(caps.get(1).unwrap().as_str())?;
+            let timelineid = ZTimelineId::from_str(caps.get(2).unwrap().as_str())?;
+            let timeline = tenant_mgr::get_timeline_for_tenant_load(tenantid, timelineid)
+                .context("Cannot load local timeline")?;
+
+            let xid: TransactionId =
+                TransactionId::from_str_radix(caps.get(3).unwrap().as_str(), 10)?;
+
+            pgb.write_message_noflush(&BeMessage::RowDescription(&[RowDescriptor::int8_col(
+                b"lsn",
+            )]))?;
+            if let Some((_timeline_id, lsn)) = timeline.lookup_lsn(None, Some(xid))? {
+                pgb.write_message_noflush(&BeMessage::DataRow(&[Some(
+                    lsn.0.to_string().as_bytes(),
+                )]))?;
+            }
```

Comment:
Actually we do not have enough information:
please notice that XIDs are assigned at start of transaction, and LSN/timestamp - at the end.
So the fact that XID=100 is absent in the database and XID=99 is present may mean two different things:
1. This transaction was not WAL logged at all (transaction was just assigned XID but failed to do anything else)
2. Information about this transaction was garbage collected (and transaction with XID=99 may be not yet collected because has more recent LSN).

Reply:
Good point. Let's leave it for now.